Fixes #30394 - allow non-admins deal with untaxed filters

ezr-ondrej opened the pull request:
Prior this non-admin user would have to have assigned Role without
taxonomies (global role) to be able to manipulate filters.
This allows manipulating Filters to any User with Filter perms.

Filters with taxonomies mean they apply to taxonomy. But given they have
taxonomies relations, they are expected to be taxable in our permission
model. All taxable resources have to have the same taxonomies as Filter
have.

Some filters doesn't have taxonomies as their underlying resource
doesn't have taxonomies. That mean they were unable to be touched by
non-admins prior this patch.

This also renames the taxable checks to express beter what they mean.

ezr-ondrej committed:
Refs #30394 - rename taxable checks

This renames the taxable check methods on Filter to better express
what they mean. We want to know it the resource is taxable, not if it
has search on taxonomy.

ezr-ondrej commented:
Alternative approach would be checking the role's taxonomies for Filter manipulation, but I believe this is what we want.

ares commented on app/models/filter.rb:
```diff
@@ -169,10 +173,10 @@ def build_taxonomy_search
+    locs = '' unless resource_taxable_by_location?
 
-    if organizations.empty? && locations.empty?
+    if organizations.blank? && locations.blank?
```
can this ever be nil? even `''` is empty

ekohl replied:
I don't think so. `build_taxonomy_search_string` guarantees to return a string.

ezr-ondrej replied:
I've rolled this back, these are not `org` and `loc`, but `organizations` and `locations`.

ares commented on app/services/authorizer.rb:
```diff
@@ -17,10 +17,12 @@ def can?(permission, subject = nil, cache = true)
-      find_collection(subject.class, :permission => permission).
-        where(:id => subject.id).any?
+      col = find_collection(subject.class, :permission => permission)
+      Foreman::Logging.logger('permissions').debug col.collect(&:id)
```
perhaps use a block for the message so `col.collect(&:id)` would only be evaluated if the logger is allowed and debug level set (performance nit)

ezr-ondrej replied:
I've just removed that.

ezr-ondrej commented on app/models/filter.rb:
```diff
@@ -169,10 +173,10 @@ def build_taxonomy_search
+    locs = '' unless resource_taxable_by_location?
 
-    if organizations.empty? && locations.empty?
+    if organizations.blank? && locations.blank?
```
this is wrong

ezr-ondrej commented on app/services/authorizer.rb:
```diff
@@ -17,10 +17,12 @@ def can?(permission, subject = nil, cache = true)
 
-      find_collection(subject.class, :permission => permission).
-        where(:id => subject.id).any?
+      col = find_collection(subject.class, :permission => permission)
+      Foreman::Logging.logger('permissions').debug col.collect(&:id)
```
Please remove this

ekohl commented on app/models/filter.rb:
```diff
@@ -169,10 +173,10 @@ def build_taxonomy_search
 
-    orgs = [] if !granular? || !resource_class.allows_organization_filtering?
-    locs = [] if !granular? || !resource_class.allows_location_filtering?
+    orgs = '' unless resource_taxable_by_organization?
```
If it's not taxable, how expensive is the call to `build_taxonomy_search_string` then? Should it rather be
```ruby
orgs = if resource_taxable_by_organization?
  build_taxonomy_search_string('organization')
else
  ''
end
```

(same for locs)

ezr-ondrej replied:
probably not significant, but even the code looks cleaner so :+1:

ekohl commented on app/models/filter.rb:
```diff
@@ -123,15 +127,15 @@ def granular?
-    allows_organization_filtering? || allows_location_filtering?
+  def resource_taxable?
+    resource_taxable_by_organization? || resource_taxable_by_location?
   end
 
-  def allows_organization_filtering?
+  def resource_taxable_by_organization?
     granular? && resource_class.allows_organization_filtering?
   end
 
-  def allows_location_filtering?
+  def resource_taxable_by_location?
     granular? && resource_class.allows_location_filtering?
   end
```
This is a public API. Is it used by any plugins?

ares replied:
Public method of a class is IMHO not a public API.

This method is defined in Authorizable module and this specific class overrides it's behavior. The method seems to be used e.g. in `app/views/filters/_form.html.erb` to decide what to show in the form. This way, it would start showing the form part for everything, ignoring the resource of the filter. If we decide to refactor to avoid the meaning misunderstanding, we need to modify it there too. We also use these methods in audit extensions, where they IMHO need to remain as is. The idea is, the filter taxonomies are delegated to the resource class of the filter.

ekohl replied:
Even if we don't consider it a public API, I am still worried about a plugin using it. A quick grep might be sufficient to check.

ezr-ondrej replied:
It is not used in any plugin in theforeman org and it should not be, I'm keeping the method from Taxonomix used by Authorizer intact, these are just methods on `Filter`.

ares reviewed (changes requested):
As discussed, please split the actual fix from the refactoring to two commits, otherwise makes sense. Left one nit re `empty?` vs `blank?`. Also the logging seems to be debuggig leftover :-)

ezr-ondrej committed:
Fixes #30394 - allow non-admins deal with untaxed filters

Prior this non-admin user would have to have assigned Role without
taxonomies (global role) to be able to manipulate filters.
This allows manipulating Filters to any User with Filter perms.

Filters with taxonomies mean they apply to taxonomy. But given they have
taxonomies relations, they are expected to be taxable in our permission
model. All taxable resources have to have the same taxonomies as Filter
have.

Some filters doesn't have taxonomies as their underlying resource
doesn't have taxonomies. That mean they were unable to be touched by
non-admins prior this patch.

This also drops current taxonomy relations in migration and force flip
the `Override` flag to false for Filter resource filters.

ezr-ondrej commented:
I've split this into two commits to have the rename method separate.

ekohl commented on app/models/filter.rb:
```diff
@@ -158,19 +162,24 @@ def enforce_inherited_taxonomies
-
-    orgs = [] if !granular? || !resource_class.allows_organization_filtering?
-    locs = [] if !granular? || !resource_class.allows_location_filtering?
+    orgs = if resource_taxable_by_organization?
+             build_taxonomy_search_string('organization')
+           else
+             ''
+           end
+    locs = if resource_taxable_by_location?
+             build_taxonomy_search_string('location')
+           else
+             ''
+           end
```
Now that I see it here, should `build_taxonomy_search_string` instead return `''` if it's not taxable?

ezr-ondrej replied:
looks much cleaner now :)

ezr-ondrej commented:
I've fixed tests, so I believe it is ready :)

tbrisker reviewed (dismissed):
LGTM, looks like all concerns have been addressed. any other comments @ares @ekohl ?

ares reviewed (commented):
I was just about to test this. Code looks good now, thanks for splitting to two commits, please keep that when merging. I haven't tested, let me know if you need hand with that. Otherwise feel free to merge.

ekohl commented:
No objections from my side.

tbrisker commented:
@ares if you have the reproducer ready for testing that would be helpful :)

ares commented:
This seems to cause problems when there is already a filter with Filter permissions that's scoped to the org/loc. From now on, it's invalid and I don't have a way to fix that from UI, because I no longer see the org/loc tabs for this filter. We may need a migration for this.

Otherwise works as expected.

ezr-ondrej commented:
Thanks @ares! Great catch! If you could verify the migration fixes the issue for you, it would be great :)

tbrisker commented on db/migrate/20210609093404_drop_override_taxonomies_from_filter.rb:
```diff
@@ -0,0 +1,8 @@
+class DropOverrideTaxonomiesFromFilter < ActiveRecord::Migration[6.0]
+  def up
+    perms = Permission.where(name: %w[view_filters create_filters edit_filters destroy_filters])
+    ids = Filter.joins(:filterings).where(filterings: { permission_id: perms }).pluck(:id).uniq
+    Filter.where(id: ids).update_all(override: false)
+    TaxableTaxonomy.where(taxable_type: 'Filter', taxable_id: ids).destroy_all
```
Not crucial, but how about:
```suggestion
    filters = Filter.joins(:filterings).where(filterings: { permission_id: perms })
    filters.update_columns(override: false)
    TaxableTaxonomy.where(taxable: filters).delete_all
```

ezr-ondrej replied:
I do not trust Rails polymorphic association with subqueries, but might be just because of historically they didn't work well, they might do work great now :shrug: 
Lets try

ezr-ondrej replied:
Works nice - at least in this case :)

tbrisker reviewed (approved):
LGTM, any other comments @ares ?

ares reviewed (approved):
Thanks, the migration passes. Merging!

tbrisker commented:
Cherry pick in GH-8616
